### ISSUE TITLE
Intersection of objects broken on 0.74

### ISSUE THREAD

**Issue by ellsclytn** [facebook/flow#6468]
[Try Flow](https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6gFSgCGAzqAEoCmRAxnpFrKqAORbV0sDciiIoAWSLoAIshLoiOGgAsAKrAAKjdCQA0oSNlCUAHkQzRKiGrAB2JPCUqmzAEyJYAnstirQAXlABvRKFAkTJQ4MshmAOYAotDWAFygAMyIAL48iDhO6JSgru5evv5mBpTxllhh4SmgAGSgGVmwkAE25g7OuSRptpagcqERnqAAFN6gRaiUGoETIRXR1qDJ8R0AlJ4AfMN+oAA8dsgAbuvb-t7jlMk+08H9UTEX2zvA+0eIK2k7fRVjxR4soSwAkFZhF5pQPN4ACyXYDrLhAA)

Flick between 0.74/0.73 to see the breakage.

**Problem**: Intersections of object types does not work for functional components in 0.74. Attempting to do so results in `Cannot create ComponentName element because property prop1 is missing in typeof intersectedProps [1] but exists in primaryProps [2].`

In my example, I'm using a typeof operator to get the object type. However, you can use a standard object type definition instead, and you'll get the same result.

**Comment by idiostruct** [comment]
0.74 does seem to introduce a regression. A workaround is to spread object types (undocumented as far as I can tell):

```typescript
type Props = {
  name: string,
  ...typeof secondaryProps
};
```

**Comment by Awpteamoose** [comment]
Object spread has a different resulting type so it's not really a workaround.

**Comment by jcready** [comment]
Intersecting with an exact object type is essentially pointless because an exact object means that no other properties can exist on the object except what is specified. The type `{name: string} & {|somethingElse: number|}` is impossible to satisfy since any object satisfying the first part of the intersection will not satisfy the second part. The `name` property from the first part cannot exist in the `{|somethingElse: number|}` exact object type.

**Comment by ellsclytn** [comment]
This then begs the question of how one is meant to merge properties of two object types.

Spread _almost_ did the trick, except that it decided to make all the properties of the spread object optional.

**Comment by jcready** [comment]
```js
type Inexact<T> = $NonMaybeType<{...T}>;

type Props = {
  name: string
} & Inexact<typeof secondaryProps>;
```

**Comment by jcready** [comment]
> Spread _almost_ did the trick, except that it decided to make all the properties of the spread object optional.

@Awpteamoose @ellsclytn I'm not sure what type you're looking at, but spreading an exact object type like in your example produces the type I would expect (no optional properties):

![image](https://user-images.githubusercontent.com/358876/41601924-9d82e7ac-73a7-11e8-917d-2c0c94207a41.png)

Flow will only make the properties of the spread object type optional **IF** the object type you're spreading is not exact. For example:

```ts
type SecondaryProps = { somethingElse: number };
type Props = {
    name: string,
    ...SecondaryProps
};
/* type-at-pos: {name: mixed, somethingElse?: number}
                       ^ now mixed         ^ now optional
```

If you want to ensure that spreading does not cause properties of inexact object types to become optional and existing properties to become `mixed` then you can simply spread the `$Exact<T>` version of them:

```ts
type SecondaryProps = { somethingElse: number };
type Props = {
    name: string,
    ...$Exact<SecondaryProps>
};
/* type-at-pos: {name: string, somethingElse: number}
                       ^ not mixed          ^ not optional
```

**Comment by goodmind** [comment]
Was fixed in Flow v0.81.0
